core/services/relay/evm/chain_reader.go

```diff
@@ -30,123 +26,98 @@ type ChainReaderService interface {
-	parsed := &parsedTypes{
-		encoderDefs: map[string]*codecEntry{},
-		decoderDefs: map[string]*codecEntry{},
+func NewChainReaderService(lggr logger.Logger, lp logpoller.LogPoller, chain legacyevm.Chain, config types.ChainReaderConfig) (ChainReaderService, error) {
+	cr := &chainReader{
+		lggr:     lggr.Named("ChainReader"),
+		lp:       lp,
+		client:   chain.Client(),
+		bindings: map[string]map[string]binding{},
+		parsed:   &parsedTypes{encoderDefs: map[string]*codecEntry{}, decoderDefs: map[string]*codecEntry{}},
+	}
+
+	var err error
+	if err = cr.init(config.ChainContractReaders); err != nil {
+		return nil, err
 	}
 
-	if err := addTypes(config.ChainContractReaders, b, parsed); err != nil {
+	if cr.codec, err = cr.parsed.toCodec(); err != nil {
 		return nil, err
 	}
 
-	c, err := parsed.toCodec()
+	err = cr.bindings.ForEach(func(b binding) error {
+		b.SetCodec(cr.codec)
+		return nil
+	})
 
-	return &chainReader{
-		lggr:     lggr.Named("ChainReader"),
-		lp:       lp,
-		codec:    c,
-		client:   chain.Client(),
-		bindings: b,
-	}, err
+	return cr, err
 }
 
 func (cr *chainReader) Name() string { return cr.lggr.Name() }
 
 var _ commontypes.ContractTypeProvider = &chainReader{}
 
 func (cr *chainReader) GetLatestValue(ctx context.Context, contractName, method string, params any, returnVal any) error {
-	ae, err := cr.bindings.getBinding(contractName, method, false)
+	b, err := cr.bindings.GetReadBinding(contractName, method)
 	if err != nil {
 		return err
 	}
 
-	if ae.evt == nil {
-		return cr.getLatestValueFromContract(ctx, contractName, method, params, returnVal)
-	}
-
-	return cr.getLatestValueFromLogPoller(ctx, contractName, method, *ae.evt, returnVal)
-}
-
-func (cr *chainReader) getLatestValueFromLogPoller(ctx context.Context, contractName, method string, hash common.Hash, returnVal any) error {
-	ae, err := cr.bindings.getBinding(contractName, method, false)
+	bytes, err := b.GetLatestValue(ctx, params)
 	if err != nil {
 		return err
 	}
-
-	log, err := cr.lp.LatestLogByEventSigWithConfs(hash, ae.addr, logpoller.Finalized)
-	if err != nil {
-		errStr := err.Error()
-		if strings.Contains(errStr, "not found") || strings.Contains(errStr, "noRows") {
-			return fmt.Errorf("%w: %w", commontypes.ErrNotFound, err)
-		}
-		return fmt.Errorf("%w: %w", commontypes.ErrInternal, err)
-	}
-	return cr.codec.Decode(ctx, log.Data, returnVal, wrapItemType(contractName, method, false))
+	return cr.codec.Decode(ctx, bytes, returnVal, wrapItemType(contractName, method, false))
 }
 
-func (cr *chainReader) getLatestValueFromContract(ctx context.Context, contractName, method string, params any, returnVal any) error {
-	data, err := cr.codec.Encode(ctx, params, wrapItemType(contractName, method, true))
-	if err != nil {
-		return err
-	}
+func (cr *chainReader) Bind(ctx context.Context, bindings []commontypes.BoundContract) error {
```

Comment:
context is unused here

Reply:
I'll do do _.  The interface needs it for LOOP because it makes a network call.